src/mock/server_options.go

```diff
@@ -132,3 +133,34 @@ func WithRequestMatchPagesEnterprise(
 
 	return WithRequestMatchPages(ep, pages...)
 }
+
+// rateLimitMiddleware enforces a rate limit using [golang.org/x/time/rate].
+// rps is the number of requests per second allowed by the rate limiter.
+// Higher burst values allow more calls to happen at once.
+// A zero value for burst will not allow any events, unless rps == [rate.Inf].
+// This middleware is intended to be used with [github.com/gorilla/mux].
+func rateLimitMiddleware(rps float64, burst int) func(next http.Handler) http.Handler {
+	limiter := rate.NewLimiter(rate.Limit(rps), burst)
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !limiter.Allow() {
+				// These values are based on this bit of logic within [github.com/google/go-github]:
+				// https://github.com/google/go-github/blob/5e25c5c215b3d21991d17447fba2e9d13a875159/github/github.go#L1243
+				w.Header().Set("X-RateLimit-Remaining", "0")
+				w.WriteHeader(http.StatusForbidden)
+
+				return
+			}
+
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+// WithRateLimit enforces a rate limit on the mocked [http.Client].
+func WithRateLimit(rps float64, burst int) MockBackendOption {
```

Comment:
I think we can start with this. I would just ask you to put some words here saying this is an alpha feature and future changes might break compatibility until a stable version is released - or something similar.

Reply:
No worries, will do, and thanks for the review 👍

Reply:
b49213c5bfe37ba72c693686a06773e1bf42c16e should resolve.